File: public/sdk/serla.ts
/**
 * Serla Analytics SDK for TypeScript/JavaScript
 * Simple, privacy-focused analytics
 *
 * Methods:
 *   - track(name, ...)  - Queue an event for batched sending (efficient for high volume)
 *   - send(name, ...)   - Send a single event immediately (for critical events)
 *   - identify(id, props) - Identify a user
 *   - flush()           - Force send all queued events
 *   - reset()           - Clear user identity and session
 *   - getSessionId()    - Get current session ID
 *   - getUserId()       - Get current user ID
 *   - setDebug(enabled) - Enable/disable debug mode
 *
 * Usage:
 *   const serla = new Serla('sk_live_...');
 *   serla.track('page_view', { distinctId: 'user_123' });  // Batched
 *   await serla.send('purchase', { properties: { amount: 99.99 } });  // Immediate
 */

interface TrackOptions {
  distinctId?: string;
  sessionId?: string;
  properties?: Record<string, unknown>;
  timestamp?: string;
  pageUrl?: string;
  pagePath?: string;
  pageTitle?: string;
  referrer?: string;
  // Click tracking fields
  clickX?: number;
  clickY?: number;
  elementSelector?: string;
}

interface IdentifyOptions {
  [key: string]: unknown;
}

interface SerlaConfig {
  endpoint?: string;
  flushInterval?: number;
  maxBatchSize?: number;
  batchSize?: number; // Alias for maxBatchSize
  debug?: boolean;
  autoTrackPageViews?: boolean;
  autoTrackClicks?: boolean;
  respectDoNotTrack?: boolean;
  sessionTimeout?: number; // In minutes
}

interface SendResponse {
  success: boolean;
  eventId: string;
  sessionId: string;
}

interface SessionData {
  sessionId: string;
  userId: string | null;
  lastActivity: number;
}

interface SerlaError extends Error {
  status?: number;
}

export class Serla {
  private apiKey: string;
  private endpoint: string;
  private queue: Array<{ name: string; options: TrackOptions }> = [];
  private flushInterval: number;
  private maxBatchSize: number;
  private flushTimer: ReturnType<typeof setTimeout> | null = null;
  private debug: boolean;
  private autoTrackPageViews: boolean;
  private autoTrackClicks: boolean;
  private respectDoNotTrack: boolean;
  private sessionTimeout: number;
  private disabled: boolean = false;
  private userId: string | null = null;
  private sessionId: string | null = null;
  private lastActivityTime: number = Date.now();

  constructor(apiKey: string, config: SerlaConfig = {}) {
    this.apiKey = apiKey;
    this.endpoint = config.endpoint || 'https://serla.dev/api/v1';
    this.flushInterval = config.flushInterval || 5000;
    this.maxBatchSize = config.maxBatchSize || config.batchSize || 10;
    this.debug = config.debug || false;
    this.autoTrackPageViews = config.autoTrackPageViews || false;
    this.autoTrackClicks = config.autoTrackClicks || false;
    this.respectDoNotTrack = config.respectDoNotTrack !== false; // Default true
    this.sessionTimeout = (config.sessionTimeout || 30) * 60 * 1000; // Convert minutes to ms

    // Check Do Not Track
    if (this.respectDoNotTrack && this.isDoNotTrackEnabled()) {
      this.log('Do Not Track is enabled, tracking disabled');
      this.disabled = true;
      return;
    }

    // Initialize session
    this.initSession();

    // Auto-track page views if enabled
    if (this.autoTrackPageViews && typeof window !== 'undefined') {
      this.setupAutoPageTracking();
    }

    // Auto-track clicks if enabled
    if (this.autoTrackClicks && typeof document !== 'undefined') {
      this.setupAutoClickTracking();
    }

    // Flush on page unload using sendBeacon for reliability
    if (typeof window !== 'undefined') {
      const flushOnUnload = () => this.flushSync();
      window.addEventListener('beforeunload', flushOnUnload);
      window.addEventListener('pagehide', flushOnUnload);
      // Also handle visibility change for mobile browsers
      document.addEventListener('visibilitychange', () => {
        if (document.visibilityState === 'hidden') {
          this.flushSync();
        }
      });
    }
  }

  private log(...args: unknown[]): void {
    if (this.debug) {
      console.log('[Serla]', ...args);
    }
  }

  private isDoNotTrackEnabled(): boolean {
    if (typeof navigator === 'undefined') return false;
    return navigator.doNotTrack === '1' ||
           navigator.doNotTrack === 'yes' ||
           // eslint-disable-next-line @typescript-eslint/no-explicit-any
           (typeof window !== 'undefined' && (window as any).doNotTrack === '1');
  }

  private generateId(): string {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = Math.random() * 16 | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
  }

  private initSession(): void {
    // Try to restore session from storage
    if (typeof localStorage !== 'undefined') {
      const stored = localStorage.getItem('serla_session');
      if (stored) {
        try {
          const data: SessionData = JSON.parse(stored);
          const now = Date.now();
          // Check if session is still valid
          if (data.lastActivity && (now - data.lastActivity) < this.sessionTimeout) {
            this.sessionId = data.sessionId;
            this.userId = data.userId;
            this.lastActivityTime = data.lastActivity;
            this.log('Restored session:', this.sessionId);
            return;
          }
        } catch {
          // Invalid stored data, create new session
        }
      }
    }

    // Create new session
    this.sessionId = 'sess_' + this.generateId();
    this.log('Created new session:', this.sessionId);
    this.saveSession();
  }

  private saveSession(): void {
    if (typeof localStorage !== 'undefined') {
      const data: SessionData = {
        sessionId: this.sessionId!,
        userId: this.userId,
        lastActivity: this.lastActivityTime,
      };
      localStorage.setItem('serla_session', JSON.stringify(data));
    }
  }

  private updateActivity(): void {
    const now = Date.now();
    // Check if session has timed out
    if ((now - this.lastActivityTime) >= this.sessionTimeout) {
      this.log('Session timed out, creating new session');
      this.sessionId = 'sess_' + this.generateId();
    }
    this.lastActivityTime = now;
    this.saveSession();
  }

  private setupAutoPageTracking(): void {
    // Track initial page view
    this.trackPageView();

    // Track on history changes (SPA navigation)
    const originalPushState = history.pushState;
    const originalReplaceState = history.replaceState;

    history.pushState = (...args: Parameters<typeof history.pushState>) => {
      originalPushState.apply(history, args);
      this.trackPageView();
    };

    history.replaceState = (...args: Parameters<typeof history.replaceState>) => {
      originalReplaceState.apply(history, args);
      this.trackPageView();
    };

    window.addEventListener('popstate', () => {
      this.trackPageView();
    });
  }

  private setupAutoClickTracking(): void {
    document.addEventListener('click', (e: MouseEvent) => {
      if (this.disabled) return;

      const target = e.target;
      if (!target || !(target instanceof Element)) return;

      // Find the nearest clickable element (button, link, etc.)
      const clickable = target.closest('a, button, [role="button"], input[type="submit"], input[type="button"]');
      const element = clickable || target;

      this.track('$click', {
        clickX: Math.round(e.clientX),
        clickY: Math.round(e.clientY),
        elementSelector: this.getSelector(element as HTMLElement),
        properties: {
          tagName: element.tagName.toLowerCase(),
          text: element.textContent?.slice(0, 100)?.trim() || undefined,
          href: (element as HTMLAnchorElement).href || undefined,
        }
      });
    }, { capture: true });
  }

  setDebug(enabled: boolean): void {
    this.debug = enabled;
    this.log('Debug mode', enabled ? 'enabled' : 'disabled');
  }

  getSessionId(): string | null {
    return this.sessionId;
  }

  getUserId(): string | null {
    return this.userId;
  }

  reset(): void {
    this.log('Resetting identity and session');
    this.userId = null;
    this.sessionId = 'sess_' + this.generateId();
    this.lastActivityTime = Date.now();
    this.saveSession();
    this.log('New session:', this.sessionId);
  }

  async track(name: string, options: TrackOptions = {}): Promise<void> {
    if (this.disabled) return;

    this.updateActivity();
    this.log('Tracking event:', name, options);

    this.queue.push({ name, options });

    if (this.queue.length >= this.maxBatchSize) {
      await this.flush();
    } else {
      this.scheduleFlush();
    }
  }

  async send(name: string, options: TrackOptions = {}): Promise<SendResponse> {
    if (this.disabled) {
      return { success: false, eventId: '', sessionId: '' } as SendResponse;
    }

    this.updateActivity();
    this.log('Sending event immediately:', name, options);

    const event: Record<string, unknown> = {
      name,
      distinctId: options.distinctId || this.userId,
      sessionId: options.sessionId || this.sessionId,
      properties: options.properties || {},
      pageUrl: options.pageUrl || (typeof window !== 'undefined' ? window.location.href : undefined),
      pagePath: options.pagePath || (typeof window !== 'undefined' ? window.location.pathname : undefined),
      pageTitle: options.pageTitle || (typeof document !== 'undefined' ? document.title : undefined),
      referrer: options.referrer || (typeof document !== 'undefined' ? document.referrer : undefined),
      clickX: options.clickX,
      clickY: options.clickY,
      elementSelector: options.elementSelector,
    };

    if (options.timestamp) {
      event.timestamp = options.timestamp;
    }

    const response = await fetch(`${this.endpoint}/events`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(event),
    });

    if (!response.ok) {
      const error: SerlaError = new Error(`Send failed: ${response.statusText}`);
      error.status = response.status;
      this.log('Send failed:', error);
      throw error;
    }

    const result = await response.json();
    this.log('Send successful:', result);
    return result;
  }

  async identify(distinctId: string, properties: IdentifyOptions = {}): Promise<void> {
    if (this.disabled) return;

    this.log('Identifying user:', distinctId, properties);
    this.userId = distinctId;
    this.saveSession();

    const response = await fetch(`${this.endpoint}/identify`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ distinctId, properties }),
    });

    if (!response.ok) {
      const error: SerlaError = new Error(`Identify failed: ${response.statusText}`);
      error.status = response.status;
      this.log('Identify failed:', error);
      throw error;
    }

    this.log('Identify successful');
  }

  private scheduleFlush(): void {
    if (this.flushTimer) return;

    this.flushTimer = setTimeout(() => {
      this.flush();
    }, this.flushInterval);
  }

  // Synchronous flush using sendBeacon for page unload scenarios
  private flushSync(): void {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }

    if (this.queue.length === 0) return;

    const events = this.queue.splice(0, this.maxBatchSize);
    this.log('Flushing (sync)', events.length, 'events');

    const payload = {
      events: events.map(({ name, options }) => ({
        name,
        distinctId: options.distinctId || this.userId,
        sessionId: options.sessionId || this.sessionId,
        properties: options.properties || {},
        timestamp: options.timestamp,
        pageUrl: options.pageUrl || (typeof window !== 'undefined' ? window.location.href : undefined),
        pagePath: options.pagePath || (typeof window !== 'undefined' ? window.location.pathname : undefined),
        pageTitle: options.pageTitle || (typeof document !== 'undefined' ? document.title : undefined),
        referrer: options.referrer || (typeof document !== 'undefined' ? document.referrer : undefined),
        clickX: options.clickX,
        clickY: options.clickY,
        elementSelector: options.elementSelector,
      })),
    };

    // Use sendBeacon for reliable delivery during page unload
    if (typeof navigator !== 'undefined' && navigator.sendBeacon) {
      const blob = new Blob([JSON.stringify(payload)], { type: 'application/json' });
      const url = `${this.endpoint}/events/batch?api_key=${this.apiKey}`;
      const sent = navigator.sendBeacon(url, blob);
      if (sent) {
        this.log('Flush (sendBeacon) successful');
      } else {
        // sendBeacon failed, re-queue events
        this.queue.unshift(...events);
        this.log('Flush (sendBeacon) failed, events re-queued');
      }
    } else {
      // Fallback: re-queue and hope for next opportunity
      this.queue.unshift(...events);
      this.log('sendBeacon not available, events re-queued');
    }
  }

  async flush(): Promise<void> {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }

    if (this.queue.length === 0) return;

    const events = this.queue.splice(0, this.maxBatchSize);
    this.log('Flushing', events.length, 'events');

    const payload = {
      events: events.map(({ name, options }) => ({
        name,
        distinctId: options.distinctId || this.userId,
        sessionId: options.sessionId || this.sessionId,
        properties: options.properties || {},
        timestamp: options.timestamp,
        pageUrl: options.pageUrl || (typeof window !== 'undefined' ? window.location.href : undefined),
        pagePath: options.pagePath || (typeof window !== 'undefined' ? window.location.pathname : undefined),
        pageTitle: options.pageTitle || (typeof document !== 'undefined' ? document.title : undefined),
        referrer: options.referrer || (typeof document !== 'undefined' ? document.referrer : undefined),
        // Click tracking fields
        clickX: options.clickX,
        clickY: options.clickY,
        elementSelector: options.elementSelector,
      })),
    };

    try {
      const response = await fetch(`${this.endpoint}/events/batch`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        // Re-queue failed events
        this.queue.unshift(...events);
        const error: SerlaError = new Error(`Batch failed: ${response.statusText}`);
        error.status = response.status;
        this.log('Flush failed:', error);
        throw error;
      }

      this.log('Flush successful');
    } catch (error) {
      // Re-queue on network failure
      this.queue.unshift(...events);
      throw error;
    }
  }

  // Auto-track page views
  trackPageView(options: Omit<TrackOptions, 'pageUrl' | 'pagePath' | 'pageTitle'> = {}): void {
    if (this.disabled) return;

    this.track('$pageview', {
      ...options,
      pageUrl: typeof window !== 'undefined' ? window.location.href : undefined,
      pagePath: typeof window !== 'undefined' ? window.location.pathname : undefined,
      pageTitle: typeof document !== 'undefined' ? document.title : undefined,
    });
  }

  // Track click with coordinates
  trackClick(element: HTMLElement, options: TrackOptions = {}): void {
    if (this.disabled) return;

    const rect = element.getBoundingClientRect();
    this.track('$click', {
      ...options,
      clickX: Math.round(rect.left + rect.width / 2),
      clickY: Math.round(rect.top + rect.height / 2),
      elementSelector: this.getSelector(element),
    });
  }

  private getSelector(element: HTMLElement): string {
    if (element.id) return `#${element.id}`;
    if (element.className) return `.${element.className.split(' ').join('.')}`;
    return element.tagName.toLowerCase();
  }
}

export default Serla;

File: public/sdk/serla.py
"""
Serla Analytics SDK for Python
Simple, privacy-focused analytics

Supports both synchronous and asynchronous usage.

Methods:
    - track(name, ...) - Queue an event for batched sending (efficient for high volume)
    - send(name, ...)  - Send a single event immediately (for critical events)
    - identify(distinct_id, properties) - Identify a user
    - flush() - Force send all queued events
    - reset() - Clear user identity and session
    - get_session_id() - Get current session ID
    - get_user_id() - Get current user ID
    - set_debug(enabled) - Enable/disable debug mode

Sync Usage:
    from serla import Serla

    serla = Serla('sk_live_...', debug=True)
    serla.track('page_view', distinct_id='user_123')  # Batched
    serla.send('purchase', distinct_id='user_123', properties={'amount': 99.99})  # Immediate
    serla.flush()  # Send remaining batched events

Async Usage:
    from serla import AsyncSerla

    async with AsyncSerla('sk_live_...') as serla:
        await serla.track('page_view', distinct_id='user_123')  # Batched
        await serla.send('purchase', properties={'amount': 99.99})  # Immediate
"""

import json
import asyncio
import threading
import uuid
from typing import Optional, Dict, Any, List
from urllib.request import Request, urlopen
from urllib.error import URLError, HTTPError
from datetime import datetime

# Check if aiohttp is available for async support
try:
    import aiohttp
    HAS_AIOHTTP = True
except ImportError:
    HAS_AIOHTTP = False


def _generate_id() -> str:
    """Generate a UUID-based ID."""
    return str(uuid.uuid4())


class Serla:
    """Synchronous Serla client with automatic batching."""

    def __init__(
        self,
        api_key: str,
        endpoint: str = "https://serla.dev/api/v1",
        flush_interval: float = 5.0,
        max_batch_size: int = 10,
        debug: bool = False,
    ):
        self.api_key = api_key
        self.endpoint = endpoint
        self.flush_interval = flush_interval
        self.max_batch_size = max_batch_size
        self._debug = debug
        self._queue: List[Dict[str, Any]] = []
        self._lock = threading.Lock()
        self._timer: Optional[threading.Timer] = None
        self._user_id: Optional[str] = None
        self._session_id: str = "sess_" + _generate_id()

    def _log(self, *args) -> None:
        """Log debug messages."""
        if self._debug:
            print("[Serla]", *args)

    def set_debug(self, enabled: bool) -> None:
        """Enable or disable debug mode."""
        self._debug = enabled
        self._log("Debug mode", "enabled" if enabled else "disabled")

    def get_session_id(self) -> str:
        """Get the current session ID."""
        return self._session_id

    def get_user_id(self) -> Optional[str]:
        """Get the current user ID."""
        return self._user_id

    def reset(self) -> None:
        """Clear user identity and create a new session."""
        self._log("Resetting identity and session")
        self._user_id = None
        self._session_id = "sess_" + _generate_id()
        self._log("New session:", self._session_id)

    def track(
        self,
        name: str,
        distinct_id: Optional[str] = None,
        session_id: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None,
        timestamp: Optional[str] = None,
    ) -> None:
        """Track an event (batched)."""
        event: Dict[str, Any] = {
            "name": name,
            "properties": properties or {},
            "timestamp": timestamp or datetime.utcnow().isoformat() + "Z",
        }
        # Use provided IDs or fall back to stored IDs
        event["distinctId"] = distinct_id or self._user_id
        event["sessionId"] = session_id or self._session_id

        self._log("Tracking event:", name, event)

        with self._lock:
            self._queue.append(event)

            if len(self._queue) >= self.max_batch_size:
                self._flush()
            else:
                self._schedule_flush()

    def send(
        self,
        name: str,
        distinct_id: Optional[str] = None,
        session_id: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None,
        timestamp: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Send a single event immediately (no batching)."""
        event: Dict[str, Any] = {
            "name": name,
            "properties": properties or {},
        }
        # Use provided IDs or fall back to stored IDs
        event["distinctId"] = distinct_id or self._user_id
        event["sessionId"] = session_id or self._session_id
        if timestamp is not None:
            event["timestamp"] = timestamp

        self._log("Sending event immediately:", name, event)
        result = self._request("POST", "/events", event)
        self._log("Send successful:", result)
        return result

    def identify(self, distinct_id: str, properties: Dict[str, Any]) -> None:
        """Identify a user with properties."""
        self._log("Identifying user:", distinct_id, properties)
        self._user_id = distinct_id
        payload = {"distinctId": distinct_id, "properties": properties}
        self._request("POST", "/identify", payload)
        self._log("Identify successful")

    def flush(self) -> None:
        """Force flush all queued events."""
        with self._lock:
            self._flush()

    def _flush(self) -> None:
        """Internal flush (must hold lock)."""
        if self._timer:
            self._timer.cancel()
            self._timer = None

        if not self._queue:
            return

        events = self._queue[: self.max_batch_size]
        self._queue = self._queue[self.max_batch_size :]

        self._log("Flushing", len(events), "events")

        try:
            self._request("POST", "/events/batch", {"events": events})
            self._log("Flush successful")
        except Exception as e:
            # Re-queue failed events
            self._queue = events + self._queue
            self._log("Flush failed:", str(e))
            raise e

    def _schedule_flush(self) -> None:
        """Schedule a flush (must hold lock)."""
        if self._timer:
            return

        self._timer = threading.Timer(self.flush_interval, self._timer_flush)
        self._timer.daemon = True
        self._timer.start()

    def _timer_flush(self) -> None:
        """Flush triggered by timer."""
        with self._lock:
            self._timer = None
            self._flush()

    def _request(self, method: str, path: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Make an HTTP request to the API."""
        url = f"{self.endpoint}{path}"
        data = json.dumps(payload).encode("utf-8")

        request = Request(
            url,
            data=data,
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
            method=method,
        )

        try:
            with urlopen(request, timeout=10) as response:
                return json.loads(response.read().decode("utf-8"))
        except HTTPError as e:
            raise Exception(f"HTTP {e.code}: {e.read().decode('utf-8')}")
        except URLError as e:
            raise Exception(f"Network error: {e.reason}")

    def __del__(self):
        """Flush remaining events on destruction."""
        try:
            self.flush()
        except Exception:
            pass


class AsyncSerla:
    """Asynchronous Serla client with automatic batching.

    Requires aiohttp: pip install aiohttp

    Usage:
        async with AsyncSerla('sk_live_...') as serla:
            await serla.track('event_name', distinct_id='user_123')

    Or:
        serla = AsyncSerla('sk_live_...')
        await serla.track('event_name')
        await serla.close()
    """

    def __init__(
        self,
        api_key: str,
        endpoint: str = "https://serla.dev/api/v1",
        flush_interval: float = 5.0,
        max_batch_size: int = 10,
        debug: bool = False,
    ):
        if not HAS_AIOHTTP:
            raise ImportError("aiohttp is required for async support. Install with: pip install aiohttp")

        self.api_key = api_key
        self.endpoint = endpoint
        self.flush_interval = flush_interval
        self.max_batch_size = max_batch_size
        self._debug = debug
        self._queue: List[Dict[str, Any]] = []
        self._lock = asyncio.Lock()
        self._flush_task: Optional[asyncio.Task] = None
        self._session: Optional[aiohttp.ClientSession] = None
        self._closed = False
        self._user_id: Optional[str] = None
        self._session_id: str = "sess_" + _generate_id()

    def _log(self, *args) -> None:
        """Log debug messages."""
        if self._debug:
            print("[Serla]", *args)

    def set_debug(self, enabled: bool) -> None:
        """Enable or disable debug mode."""
        self._debug = enabled
        self._log("Debug mode", "enabled" if enabled else "disabled")

    def get_session_id(self) -> str:
        """Get the current session ID."""
        return self._session_id

    def get_user_id(self) -> Optional[str]:
        """Get the current user ID."""
        return self._user_id

    def reset(self) -> None:
        """Clear user identity and create a new session."""
        self._log("Resetting identity and session")
        self._user_id = None
        self._session_id = "sess_" + _generate_id()
        self._log("New session:", self._session_id)

    async def _get_session(self) -> aiohttp.ClientSession:
        """Get or create the aiohttp session."""
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                },
                timeout=aiohttp.ClientTimeout(total=10),
            )
        return self._session

    async def track(
        self,
        name: str,
        distinct_id: Optional[str] = None,
        session_id: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None,
        timestamp: Optional[str] = None,
    ) -> None:
        """Track an event asynchronously (batched)."""
        if self._closed:
            raise RuntimeError("Client has been closed")

        event: Dict[str, Any] = {
            "name": name,
            "properties": properties or {},
            "timestamp": timestamp or datetime.utcnow().isoformat() + "Z",
        }
        # Use provided IDs or fall back to stored IDs
        event["distinctId"] = distinct_id or self._user_id
        event["sessionId"] = session_id or self._session_id

        self._log("Tracking event:", name, event)

        async with self._lock:
            self._queue.append(event)

            if len(self._queue) >= self.max_batch_size:
                await self._flush()
            else:
                self._schedule_flush()

    async def send(
        self,
        name: str,
        distinct_id: Optional[str] = None,
        session_id: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None,
        timestamp: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Send a single event immediately (no batching)."""
        if self._closed:
            raise RuntimeError("Client has been closed")

        event: Dict[str, Any] = {
            "name": name,
            "properties": properties or {},
        }
        # Use provided IDs or fall back to stored IDs
        event["distinctId"] = distinct_id or self._user_id
        event["sessionId"] = session_id or self._session_id
        if timestamp is not None:
            event["timestamp"] = timestamp

        self._log("Sending event immediately:", name, event)
        result = await self._request("POST", "/events", event)
        self._log("Send successful:", result)
        return result

    async def identify(self, distinct_id: str, properties: Dict[str, Any]) -> None:
        """Identify a user with properties asynchronously."""
        if self._closed:
            raise RuntimeError("Client has been closed")
        self._log("Identifying user:", distinct_id, properties)
        self._user_id = distinct_id
        payload = {"distinctId": distinct_id, "properties": properties}
        await self._request("POST", "/identify", payload)
        self._log("Identify successful")

    async def flush(self) -> None:
        """Force flush all queued events asynchronously."""
        async with self._lock:
            await self._flush()

    async def _flush(self) -> None:
        """Internal flush (must hold lock)."""
        if self._flush_task and not self._flush_task.done():
            self._flush_task.cancel()
            self._flush_task = None

        if not self._queue:
            return

        events = self._queue[: self.max_batch_size]
        self._queue = self._queue[self.max_batch_size :]

        self._log("Flushing", len(events), "events")

        try:
            await self._request("POST", "/events/batch", {"events": events})
            self._log("Flush successful")
        except Exception as e:
            # Re-queue failed events
            self._queue = events + self._queue
            self._log("Flush failed:", str(e))
            raise e

    def _schedule_flush(self) -> None:
        """Schedule a flush."""
        if self._flush_task and not self._flush_task.done():
            return

        async def delayed_flush():
            await asyncio.sleep(self.flush_interval)
            async with self._lock:
                await self._flush()

        self._flush_task = asyncio.create_task(delayed_flush())

    async def _request(self, method: str, path: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Make an async HTTP request to the API."""
        url = f"{self.endpoint}{path}"
        session = await self._get_session()

        async with session.request(method, url, json=payload) as response:
            if response.status >= 400:
                text = await response.text()
                raise Exception(f"HTTP {response.status}: {text}")
            return await response.json()

    async def close(self) -> None:
        """Close the client and flush remaining events."""
        if self._closed:
            return
        self._closed = True

        try:
            await self.flush()
        except Exception:
            pass

        if self._flush_task and not self._flush_task.done():
            self._flush_task.cancel()

        if self._session and not self._session.closed:
            await self._session.close()

    async def __aenter__(self) -> "AsyncSerla":
        """Async context manager entry."""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Async context manager exit."""
        await self.close()


# Convenience functions for sync usage
_default_client: Optional[Serla] = None


def init(api_key: str, **kwargs) -> Serla:
    """Initialize the default Serla client."""
    global _default_client
    _default_client = Serla(api_key, **kwargs)
    return _default_client


def track(name: str, **kwargs) -> None:
    """Track an event using the default client (batched)."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    _default_client.track(name, **kwargs)


def send(name: str, **kwargs) -> Dict[str, Any]:
    """Send a single event immediately using the default client (no batching)."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    return _default_client.send(name, **kwargs)


def identify(distinct_id: str, properties: Dict[str, Any]) -> None:
    """Identify a user using the default client."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    _default_client.identify(distinct_id, properties)


def flush() -> None:
    """Flush events using the default client."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    _default_client.flush()


def reset() -> None:
    """Reset identity using the default client."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    _default_client.reset()


def get_session_id() -> str:
    """Get session ID using the default client."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    return _default_client.get_session_id()


def get_user_id() -> Optional[str]:
    """Get user ID using the default client."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    return _default_client.get_user_id()


def set_debug(enabled: bool) -> None:
    """Set debug mode using the default client."""
    if _default_client is None:
        raise RuntimeError("Call serla.init() first")
    _default_client.set_debug(enabled)


# Async convenience functions
_async_client: Optional[AsyncSerla] = None


async def async_init(api_key: str, **kwargs) -> AsyncSerla:
    """Initialize the default async Serla client."""
    global _async_client
    _async_client = AsyncSerla(api_key, **kwargs)
    return _async_client


async def async_track(name: str, **kwargs) -> None:
    """Track an event using the default async client (batched)."""
    if _async_client is None:
        raise RuntimeError("Call await serla.async_init() first")
    await _async_client.track(name, **kwargs)


async def async_send(name: str, **kwargs) -> Dict[str, Any]:
    """Send a single event immediately using the default async client (no batching)."""
    if _async_client is None:
        raise RuntimeError("Call await serla.async_init() first")
    return await _async_client.send(name, **kwargs)


async def async_identify(distinct_id: str, properties: Dict[str, Any]) -> None:
    """Identify a user using the default async client."""
    if _async_client is None:
        raise RuntimeError("Call await serla.async_init() first")
    await _async_client.identify(distinct_id, properties)


async def async_flush() -> None:
    """Flush events using the default async client."""
    if _async_client is None:
        raise RuntimeError("Call await serla.async_init() first")
    await _async_client.flush()


async def async_reset() -> None:
    """Reset identity using the default async client."""
    if _async_client is None:
        raise RuntimeError("Call await serla.async_init() first")
    _async_client.reset()


async def async_close() -> None:
    """Close the default async client."""
    global _async_client
    if _async_client is not None:
        await _async_client.close()
        _async_client = None
